hmacauth.go

```diff
@@ -5,35 +5,100 @@ import (
 	"crypto"
 	"crypto/hmac"
 	"encoding/base64"
+	"errors"
 	"net/http"
 	"strconv"
 	"strings"
 )
 
-var supportedAlgorithms map[string]crypto.Hash
+var supportedAlgorithms map[string]crypto.Hash = map[string]crypto.Hash{
+	"md4":       crypto.MD4,
+	"md5":       crypto.MD5,
+	"sha1":      crypto.SHA1,
+	"sha224":    crypto.SHA224,
+	"sha256":    crypto.SHA256,
+	"sha384":    crypto.SHA384,
+	"sha512":    crypto.SHA512,
+	"ripemd160": crypto.RIPEMD160,
+}
+
 var algorithmName map[crypto.Hash]string
 
 func init() {
-	supportedAlgorithms = map[string]crypto.Hash{
-		"sha1": crypto.SHA1,
-	}
-
 	algorithmName = make(map[crypto.Hash]string)
 	for name, algorithm := range supportedAlgorithms {
 		algorithmName[algorithm] = name
+		if algorithm.Available() == false {
+			delete(supportedAlgorithms, name)
+		}
 	}
 }
 
-func StringToSign(req *http.Request, headers []string) string {
+func DigestNameToCryptoHash(name string) (result crypto.Hash, err error) {
+	if result = supportedAlgorithms[name]; result == crypto.Hash(0) {
```

Comment:
Use the map key test talked about [here](https://blog.golang.org/go-maps-in-action)
```go
if _, ok := supportedAlgorithms[name]; !ok {
...
```